In [1]:
#===============================================================================
# Import libraries
#===============================================================================
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [32]:
#===============================================================================
# Load the dataset
#===============================================================================
print('loading data ...')
df = pd.read_csv('./dataset/CrowdstormingDataJuly1st.csv')
print('data is loaded.')

loading data ...
data is loaded.


## Variables description:
- playerShort: short player ID
- player: player name
- club: player club
- leagueCountry: country of player club (England, Germany, France, and Spain)
- birthday: player birthday
- height: player height (in cm)
- weight: player weight (in kg)
- position: detailed player position
- games: number of games in the player-referee dyad
- victories: victories in the player-referee dyad
- ties: ties in the player-referee dyad
- defeats: losses in the player-referee dyad
- goals: goals scored by a player in the player-referee dyad
- yellowCards: number of yellow cards player received from referee
- yellowReds: number of yellow-red cards player received from referee
- redCards: number of red cards player received from referee
- photoID: ID of player photo (if available)
- rater1: skin rating of photo by rater 1 (5-point scale ranging from “very light skin” to “very dark skin”)
- rater2: skin rating of photo by rater 2 (5-point scale ranging from “very light skin” to “very dark skin”)
- refNum: unique referee ID number (referee name removed for anonymizing purposes)
- refCountry: unique referee country ID number (country name removed for anonymizing purposes)
- meanIAT: mean implicit bias score (using the race IAT) for referee country, higher values correspond to faster white | good, black | bad associations
- nIAT: sample size for race IAT in that particular country
- seIAT: standard error for mean estimate of race IAT
- meanExp: mean explicit bias score (using a racial thermometer task) for referee country, higher values correspond to greater feelings of warmth toward whites versus blacks
- nExp: sample size for explicit bias in that particular country
- seExp: standard error for mean estimate of explicit bias measure

In [10]:
# show the first 10 columns for the first 10 players
df.iloc[:10,:10]

playerShort              player               club leagueCountry  \
0       lucas-wilchez       Lucas Wilchez      Real Zaragoza         Spain   
1          john-utaka          John Utaka    Montpellier HSC        France   
2         abdon-prats         Abdón Prats       RCD Mallorca         Spain   
3          pablo-mari          Pablo Marí       RCD Mallorca         Spain   
4          ruben-pena          Rubén Peña    Real Valladolid         Spain   
5        aaron-hughes        Aaron Hughes          Fulham FC       England   
6  aleksandar-kolarov  Aleksandar Kolarov    Manchester City       England   
7    alexander-tettey    Alexander Tettey       Norwich City       England   
8   anders-lindegaard   Anders Lindegaard  Manchester United       England   
9        andreas-beck        Andreas Beck    1899 Hoffenheim       Germany   

     birthday  height  weight              position  games  victories  
0  31.08.1983   177.0    72.0  Attacking Midfielder      1          0  
1  08.01.1982   179.0    82.0          Right Winger      1          0  
2  17.12.1992   181.0    79.0                   NaN      1          0  
3  31.08.1993   191.0    87.0           Center Back      1          1  
4  18.07.1991   172.0    70.0      Right Midfielder      1          1  
5  08.11.1979   182.0    71.0           Center Back      1          0  
6  10.11.1985   187.0    80.0         Left Fullback      1          1  
7  04.04.1986   180.0    68.0  Defensive Midfielder      1          0  
8  13.04.1984   193.0    80.0            Goalkeeper      1          0  
9  13.03.1987   180.0    70.0        Right Fullback      1          1

In [16]:
# show columns 10 to 21 for the first 10 players
df.iloc[:10,10:22]

ties  defeats  goals  yellowCards  yellowReds  redCards    photoID  rater1  \
0     0        1      0            0           0         0  95212.jpg    0.25   
1     0        1      0            1           0         0   1663.jpg    0.75   
2     1        0      0            1           0         0        NaN     NaN   
3     0        0      0            0           0         0        NaN     NaN   
4     0        0      0            0           0         0        NaN     NaN   
5     0        1      0            0           0         0   3868.jpg    0.25   
6     0        0      0            0           0         0  47704.jpg    0.00   
7     0        1      0            0           0         0  22356.jpg    1.00   
8     1        0      0            0           0         0  16528.jpg    0.25   
9     0        0      0            0           0         0  36499.jpg    0.00   

   rater2  refNum  refCountry Alpha_3  
0    0.50       1           1     GRC  
1    0.75       2           2     ZMB  
2     NaN       3           3     ESP  
3     NaN       3           3     ESP  
4     NaN       3           3     ESP  
5    0.00       4           4     LUX  
6    0.25       4           4     LUX  
7    1.00       4           4     LUX  
8    0.25       4           4     LUX  
9    0.00       4           4     LUX

In [17]:
# show the last 6 columns for the first 10 players
df.iloc[:10,22::]

meanIAT    nIAT     seIAT   meanExp    nExp     seExp
0  0.326391   712.0  0.000564  0.396000   750.0  0.002696
1  0.203375    40.0  0.010875 -0.204082    49.0  0.061504
2  0.369894  1785.0  0.000229  0.588297  1897.0  0.001002
3  0.369894  1785.0  0.000229  0.588297  1897.0  0.001002
4  0.369894  1785.0  0.000229  0.588297  1897.0  0.001002
5  0.325185   127.0  0.003297  0.538462   130.0  0.013752
6  0.325185   127.0  0.003297  0.538462   130.0  0.013752
7  0.325185   127.0  0.003297  0.538462   130.0  0.013752
8  0.325185   127.0  0.003297  0.538462   130.0  0.013752
9  0.325185   127.0  0.003297  0.538462   130.0  0.013752

In [19]:
# the data desrption
df.describe()

C:\Users\shafieez\AppData\Local\Continuum\Anaconda3\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


height         weight          games      victories  \
count  145765.000000  143785.000000  146028.000000  146028.000000   
mean      181.935938      76.075662       2.921166       1.278344   
std         6.738726       7.140906       3.413633       1.790725   
min       161.000000      54.000000       1.000000       0.000000   
25%              NaN            NaN       1.000000       0.000000   
50%              NaN            NaN       2.000000       1.000000   
75%              NaN            NaN       3.000000       2.000000   
max       203.000000     100.000000      47.000000      29.000000   

                ties        defeats          goals    yellowCards  \
count  146028.000000  146028.000000  146028.000000  146028.000000   
mean        0.708241       0.934581       0.338058       0.385364   
std         1.116793       1.383059       0.906481       0.795333   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       1.000000       0.000000       0.000000   
75%         1.000000       1.000000       0.000000       1.000000   
max        14.000000      18.000000      23.000000      14.000000   

          yellowReds       redCards         rater1         rater2  \
count  146028.000000  146028.000000  124621.000000  124621.000000   
mean        0.011381       0.012559       0.264255       0.302862   
std         0.107931       0.112889       0.295382       0.293020   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000            NaN            NaN   
50%         0.000000       0.000000            NaN            NaN   
75%         0.000000       0.000000            NaN            NaN   
max         3.000000       2.000000       1.000000       1.000000   

              refNum     refCountry        meanIAT          nIAT  \
count  146028.000000  146028.000000  145865.000000  1.458650e+05   
mean     1534.827444      29.642842       0.346276  1.969741e+04   
std       918.736625      27.496189       0.032246  1.271262e+05   
min         1.000000       1.000000      -0.047254  2.000000e+00   
25%       641.000000       7.000000            NaN           NaN   
50%      1604.000000      21.000000            NaN           NaN   
75%      2345.000000      44.000000            NaN           NaN   
max      3147.000000     161.000000       0.573793  1.975803e+06   

              seIAT        meanExp          nExp          seExp  
count  1.458650e+05  145865.000000  1.458650e+05  145865.000000  
mean   6.310849e-04       0.452026  2.044023e+04       0.002994  
std    4.735857e-03       0.217469  1.306157e+05       0.019723  
min    2.235373e-07      -1.375000  2.000000e+00       0.000001  
25%             NaN            NaN           NaN            NaN  
50%             NaN            NaN           NaN            NaN  
75%             NaN            NaN           NaN            NaN  
max    2.862871e-01       1.800000  2.029548e+06       1.060660

As we can see from the first 10 players and data desription, there are some inconsistencies in data, which are as following:
- Some players' photo are not available. Hence, there is no information on their skin color.
- The race IAT informatios are not availbale for some regions/players.
- For some players, the height/weight informations are not given.
- The two raters disagree about player skintone.
- Some refs officiated very few games. It might be possible that they are suspended after a poor referee.

In [33]:
# First of all, we remove all players whose skin color are not available
print('The size of data before any cleaning:', df.shape)
df.dropna(subset=['photoID', 'rater1', 'rater2'], how='any', inplace=True)
print('The size of data after droping all players whose skin color are not available:', df.shape)

The size of data before any cleaning: (146028, 28)
The size of data after droping all players whose skin color are not available: (124621, 28)


So we almost loose 20000 dyads. Now let see the description of the resulting dataframe.

In [34]:
df.describe()

C:\Users\shafieez\AppData\Local\Continuum\Anaconda3\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


height         weight          games      victories  \
count  124575.000000  123868.000000  124621.000000  124621.000000   
mean      182.125330      76.368102       2.993613       1.342663   
std         6.804843       7.188595       3.580429       1.879913   
min       161.000000      54.000000       1.000000       0.000000   
25%              NaN            NaN       1.000000       0.000000   
50%              NaN            NaN       1.000000       1.000000   
75%              NaN            NaN       3.000000       2.000000   
max       203.000000     100.000000      47.000000      29.000000   

                ties        defeats          goals    yellowCards  \
count  124621.000000  124621.000000  124621.000000  124621.000000   
mean        0.714968       0.935982       0.353432       0.399170   
std         1.143873       1.422906       0.944250       0.821651   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       1.000000       0.000000       0.000000   
75%         1.000000       1.000000       0.000000       1.000000   
max        14.000000      18.000000      23.000000      14.000000   

          yellowReds       redCards         rater1         rater2  \
count  124621.000000  124621.000000  124621.000000  124621.000000   
mean        0.012061       0.012751       0.264255       0.302862   
std         0.111052       0.113901       0.295382       0.293020   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.250000       0.250000   
75%         0.000000       0.000000       0.250000       0.500000   
max         3.000000       2.000000       1.000000       1.000000   

              refNum     refCountry        meanIAT          nIAT  \
count  124621.000000  124621.000000  124468.000000  1.244680e+05   
mean     1531.111546      29.488096       0.348123  1.817187e+04   
std       916.314559      27.963752       0.032087  1.261437e+05   
min         1.000000       1.000000      -0.047254  2.000000e+00   
25%       652.000000       7.000000            NaN           NaN   
50%      1574.000000      15.000000            NaN           NaN   
75%      2332.000000      45.000000            NaN           NaN   
max      3147.000000     161.000000       0.573793  1.975803e+06   

              seIAT        meanExp          nExp          seExp  
count  1.244680e+05  124468.000000  1.244680e+05  124468.000000  
mean   6.287461e-04       0.463797  1.884147e+04       0.002988  
std    4.788647e-03       0.219206  1.296022e+05       0.019719  
min    2.235373e-07      -1.375000  2.000000e+00       0.000001  
25%             NaN            NaN           NaN            NaN  
50%             NaN            NaN           NaN            NaN  
75%             NaN            NaN           NaN            NaN  
max    2.862871e-01       1.800000  2.029548e+06       1.060660

The race IAT informatios are also crucial to make any conclusion on refs. Let first see for how many rows these informations are not available.

In [53]:
df[df.iloc[:, 22::].isnull().any(axis=1)].shape

(153, 28)

Hence, there are 153 row with missing informations in race IAT. Lets drop them.

In [59]:
df.dropna(subset=df.columns[22::], how='any', inplace=True)
print('The size of data is now', df.shape)

The size of data is now (124468, 28)


Lets consider the data with missing values for wieght and height columns.

In [64]:
print('Number of missing weight data:', sum(df['weight'].isnull()))
print('Number of missing height data:', sum(df['height'].isnull()))

Number of missing weight data: 753
Number of missing height data: 46


In [66]:
df.dropna(subset=['weight', 'height'], how='any', inplace=True)
print('The size of data after removing all NaN is', df.shape)

The size of data after removing all NaN is (123715, 28)


In [ ]:
# Save the data without NaN
df.to_csv('./dataset/players_without_nan.csv', index=False)